homeassistant/components/switcher_kis/switch.py

```diff
@@ -0,0 +1,179 @@
+
+    @property
+    def electric_current(self) -> float:
+        """Return the electric current."""
+        return cast(float, self._device_data.electric_current)
+
+    @property
+    def remaining_time(self) -> str:
+        """Return the remaining time to off command."""
+        return cast(str, self._device_data.remaining_time)
+
+    @property
+    def auto_off_set(self) -> str:
+        """Return the auto off configuration set."""
+        return cast(str, self._device_data.auto_off_set)
+
+    @property
+    def device_name(self) -> str:
+        """Return the device's name."""
+        return cast(str, self._device_data.name)
+
+    @property
+    def should_poll(self) -> bool:
+        """Return False, entity pushes its state to HA."""
+        return False
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return "{}-{}".format(
+            self._device_data.device_id, self._device_data.mac_addr)
+
+    @property
+    def is_on(self) -> bool:
+        """Return True if entity is on."""
+        from aioswitcher.consts import STATE_ON as SWITCHER_STATE_ON
+        return cast(bool, self._state == SWITCHER_STATE_ON)
+
+    @property
+    def current_power_w(self) -> int:
+        """Return the current power usage in W."""
+        return cast(int, self._device_data.power_consumption)
+
+    @property
+    def device_state_attributes(self) -> Dict:
+        """Return the optional state attributes."""
+        from aioswitcher.consts import WAITING_TEXT
+
+        attribs = {}
+
+        for prop, attr in PROPERTIES_TO_DEVICE_ATTRIBUTES.items():
+            value = getattr(self, prop)
+            if value and value is not WAITING_TEXT:
+                attribs[attr] = value
+
+        return attribs
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        from aioswitcher.consts import (STATE_OFF as SWITCHER_STATE_OFF,
+                                        STATE_ON as SWITCHER_STATE_ON)
+        return self._state in [SWITCHER_STATE_ON, SWITCHER_STATE_OFF]
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+        if not self.hass:
+            return
+        self._unsub_dispatcher = async_dispatcher_connect(
+            self.hass, SIGNAL_SWITCHER_DEVICE_UPDATE, self.async_update_data)
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Run when entity will be removed from hass."""
+        if self._unsub_dispatcher:
+            self._unsub_dispatcher()
+
+    async def async_update_data(self, device_data: Any) -> None:
+        """Update the entity data."""
+        # pylint: disable=unused-import
+        from aioswitcher.devices import SwitcherV2Device
+        # pylint: enable=unused-import
+        temp_device = device_data  # type: SwitcherV2Device
+        if temp_device:
+            if self._self_initiated:
```

Comment:
I have a little confuse about this, is `_self_initiated` really means as it's name?

Reply:
@awarecan 
First of all, thank you for your much appreciated review, I've already started working on your change requests.

In regards to your question:
The data updates from the device comes via a UDP connection approximately every 4 seconds.
The commands send to device, for example turn on/off are manual TCP requests.

When a TCP request for turning on the device is sent just before or whilst the UDP data update is being transmitted, we end up with a false outcome because according to the data update, the device is off but in HA it just been marked on.
The result is the entity being turned on and immediately being turned off (although the device is actually turned on) and in about 4 seconds it'll get the correct data update and get turned on again.

To resolve this, the `_self_initiated==True` marks the the turn on/off request was sent by HA entity, in which case the update process will skip the current data update and re-set `_self_initiated=False` so that the next update go on correctly.

Maybe the name `_self_initiated` is confusing.
Would you recommend changing it perhaps?

Reply:
I initially read it as `initialized` -- wonder if you did too @awarecan.  I think `initiated` is fine in this case.  Ideally, this would be logic handled by the library -- not inside HASS, but you can address that in the future.

Reply:
@andrewsayre, you are exactly right, I have read it as `initialized` 😄